calico_kubernetes/calico_kubernetes_cni.py

```diff
@@ -307,6 +290,297 @@ def _get_container_pid(container_id):
+        outbound_rules = [["allow"]]
+    else:
+        inbound_rules = [["allow"]]
+        outbound_rules = [["allow"]]
+
+    _log.info("Getting Policy Rules from Annotation of pod %s", pod_name)
+
+    annotations = _get_metadata(namespace, pod_name, "annotations")
+
+    # Find policy block of annotations
+    if annotations and POLICY_ANNOTATION_KEY in annotations:
+        # Remove Default Rule (Allow Namespace)
+        inbound_rules = []
+        rules = annotations[POLICY_ANNOTATION_KEY]
+
+        # Rules separated by semicolons
+        for rule in rules.split(";"):
+            args = rule.split(" ")
+
+            # Labels are declared in the annotations with the format 'label X=Y'
+            # These must be converted into format 'tag NAMESPACE_X_Y' to be parsed by calicoctl.
+            if 'label' in args:
+                # Replace arg 'label' with arg 'tag'
+                label_ind = args.index('label')
+                args[label_ind] = 'tag'
+
+                # Split given label 'key=value' into components 'key', 'value'
+                label = args[label_ind + 1]
+                key, value = label.split('=')
+
+                # Compose Calico tag out of key, value components
+                tag = _label_to_tag(key, value)
+                args[label_ind + 1] = tag
+
+            # Remove empty strings and add to rule list
+            args = filter(None, args)
+            inbound_rules.append(args)
+
+    return inbound_rules, outbound_rules
+
+
+def _label_to_tag(ns, label_key, label_value):
+    """
+    Labels are key-value pairs, tags are single strings. This function handles that translation
+    1) Concatenate key and value with '='
+    2) Prepend a pod's namespace followed by '/' if available
+    3) Escape the generated string so it is Calico compatible
+
+    :param label_key: key to label
+    :param label_value: value to given key for a label
+    :param namespace: Namespace string, input None if not available
+    :param types: (self, string, string, string)
+    :return single string tag
+    :rtype string
+    """
+    tag = '%s=%s' % (label_key, label_value)
+    tag = '%s/%s' % (ns, tag)
+    tag = _escape_chars(tag)
+    return tag
+
+
+def _get_namespace_tag(namespace):
+    """
+    Pull metadata for namespace and return it and a generated NS tag
+    """
+    ns_tag = _escape_chars('%s=%s' % ('namespace', ns))
+    return ns_tag
+
+
+def _escape_chars(self, unescaped_string):
+    """
+    Calico can only handle 3 special chars, '_.-'
+    This function uses regex sub to replace SCs with '_'
+    """
+    # Character to replace symbols
+    swap_char = '_'
+
+    # If swap_char is in string, double it.
+    unescaped_string = re.sub(swap_char, "%s%s" % (swap_char, swap_char), unescaped_string)
+
+    # Substitute all invalid chars.
+    return re.sub('[^a-zA-Z0-9\.\_\-]', swap_char, unescaped_string)
+
+
+def _get_metadata(namespace, pod_name, key):
```

Comment:
We want to minimize the number of times our plugin is hitting the API, since each access is quite costly.

We should be doing this once at startup (since we know we'll need the data) and storing it so that we can access our copy on demand without hitting the API again.